benchmark/NCM_Stress_Test_Report/NCM Stress Test Report.adoc

```diff
@@ -0,0 +1,97 @@
+= NCM Stress Test Report
+:revnumber: v1.0
+:revdate: 2022 January 28
+:author: Rio Zhu
+:email: zzhu@futurewei.com
+
+:toc: right
+:imagesdir: images
+
+== Overview
+
+This document presents the results of running on-demand stress tests against Network-Configuration-Manager(NCM), with Ignite as the cache layer of NCM. The reason of this test is to get a better understanding of how much on-demand requests can NCM handle in a certain amount of time.
+
+The following graph shows the setup and the workflow of this test.
+
+We tested NCM with different numbers of VPCs, and different number of ports inside each VPC.
+
+image::ncm_stress_test_setup.png[test_setup, 800]
+
+== Test Setup and Detailed Workflow
+
+The test setup is very similar with our previous tests on the on-demand workflow. Except that we replced one of the Alcor-Control-Agent(ACA) on one of the nodes with the newly written Dubhe Agent, a prototype written in Golang that mimics the gRPC server/client ability of ACA. Of course, for the Test Controller, changes were made in order to support the multi-vpc test scenarios.
+
+At the beginning of the test, the Test Controller first generates the GoalStateV2 messages with the number of VPCs, and number of ports inside each VPC, and it sends those GoalStateV2s to NCM.
+
+NCM, after receiving and storing these GoalStateV2s into its cache, sends down the GoalStates to the host agents, which are the Dubhe agent on Node 1 and ACA on Node 2. 
+
+With the GoalStateV2 received, NCM now understands what VPCs and ports it has and what neighbors those ports have, and it is ready for begin the stress test. The user sets the duration of the test(in seconds) and the rate of sending on-demand requests(requests/second, we can call this number X) when the Dubhe agent starts, and test will be done according to these parameters.
+
+After the test, the Dubhe agent outputs the number of requests sent, the number of GoalStateV2 it received, and the number of on-demand replies it received. If these three numbers are at the same level, we would say that NCM is able to handle X number of on-demand requests per second, with this number of VPCs and ports in its cache.
```

Comment:
Not the exact number, because when the time is up, there are usually some replies that didn't come back on time and they are missed.

I'd say that if the difference is within one or two hundreds, it is still the same level. Because, even if the on-demand rate is slightly bigger than what NCM can handle, say we set the rate to 750, but atually the NCM max QPS is 700, we will see a significant(I'll say at least 30%) drop of GoalStateV2s received and the on-demand replies received.